common/src/main/java/io/netty/util/concurrent/SimpleConcurrencyLimit.java

```diff
@@ -0,0 +1,197 @@
+ * under the License.
+ */
+package io.netty.util.concurrent;
+
+import static io.netty.util.internal.ObjectUtil.checkPositive;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import org.jetbrains.annotations.VisibleForTesting;
+
+import io.netty.util.internal.UnstableApi;
+
+/**
+ * A simple {@link ConcurrencyLimit} that limits the number of permits allowed concurrently.
+ * It is similar to {@link Semaphore} in its behavior, except that it's asynchronous and callback-driven.
+ * <pre>{@code
+ * limit = new SimpleConcurrencyLimit(42);
+ * handler = new ConcurrencyLimitHandler() {
+ *   public void permitAcquired(Runnable releaser) {
+ *     System.err.println("Acquired a permit!");
+ *     ...
+ *     // Release the acquired permit.
+ *     releaser.run();
+ *   }
+ *
+ *   public void permitAcquisitionTimedOut() {
+ *     System.err.println("Failed to acquire a permit due to timeout.");
+ *   }
+ * }
+ *
+ * limit.acquire(eventLoop, handler, 10, TimeUnit.SECONDS);
+ * }</pre>
+ */
+@UnstableApi
+public class SimpleConcurrencyLimit implements ConcurrencyLimit {
+
+    private static final AtomicIntegerFieldUpdater<SimpleConcurrencyLimit> permitsInUseUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(SimpleConcurrencyLimit.class, "permitsInUse");
+
+    private volatile int permitsInUse;
+    private final int maxPermits;
+    private final Releaser releaser = new Releaser();
+    private final Queue<PendingHandler> pendingHandlers = new ConcurrentLinkedQueue<PendingHandler>();
+
+    public SimpleConcurrencyLimit(int maxPermits) {
+        this.maxPermits = checkPositive(maxPermits, "maxPermits");
+    }
+
+    @VisibleForTesting
+    int permitsInUse() {
+        return permitsInUse;
+    }
+
+    @Override
+    public void acquire(EventExecutor executor, final ConcurrencyLimitHandler handler,
+                        long timeout, TimeUnit unit) {
+        for (;;) {
+            final int currentPermitsInUse = permitsInUse;
+            if (currentPermitsInUse == maxPermits) {
+                final PendingHandler pendingHandler = new PendingHandler(executor, handler);
+                if (timeout > 0) {
+                    final ScheduledFuture<?> timeoutFuture = executor.schedule(new Runnable() {
+                        @Override
+                        public void run() {
+                            if (pendingHandler.isTimeoutFutureDone()) {
+                                return;
+                            }
+
+                            pendingHandler.handler.permitAcquisitionTimedOut();
+                        }
+                    }, timeout, unit);
+                    pendingHandler.setTimeoutFuture(timeoutFuture);
+                }
+
+                pendingHandlers.add(pendingHandler);
+                return;
```

Comment:
We have a data-race here where all permits could have been returned, and all releasers could have been called, before we manage to add our pending handler to the queue.